### PR TITLE
+ GPUifyLoops.jl to Utilities (for MoistThermo)

### PR DESCRIPTION
**Pull request by charleskawczynski**
Trying to add GPUifyLoops to Utilities so that MoistThermodynamics can handle GPU arrays-- `exp` is called inside MoistThermodynamics

### CODE REVIEW

File: src/Utilities/test/runtests.jl [hunk 1]
```diff
@@ -128,44 +131,72 @@ end
...
-    #   true
-    # catch
-    #   false
-    # end
-    # @test L
+  @require CUDAnative="be33ccc6-a3ff-5ff2-a52e-74243cff1e17" begin
+    using .CUDAnative
+    using .CUDAnative.CUDAdrv
+
+    @testset "MoistThermo CUDA" begin
```
Comment (charleskawczynski) [inline src/Utilities/test/runtests.jl]: @vchuravy, I tried following the example you gave
```
kernel(args....)
cukernel(args...) = contextualize(kernel)(args...)
@cuda cukernel()
```
but I'm getting
```
┌ Warning: Error requiring CUDAnative from Main:
│ LoadError: ArgumentError: second argument to @cuda should be a function call
│ Stacktrace:
│  [1] @cuda(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at C:\Users\kawcz\.julia\packages\CUDAnative\Mdd3w\src\execution.jl:189
│  [2] eval at .\boot.jl:319 [inlined]
│  [3] eval(::Expr) at .\client.jl:389
│  [4] (::getfield(Main, Symbol("##9#12")))() at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:67
│  [5] err(::getfield(Main, Symbol("##9#12")), ::Module, ::String) at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:38
│  [6] #8 at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:66 [inlined]
│  [7] withpath(::getfield(Main, Symbol("##8#11")), ::String) at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:28
│  [8] #7 at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:65 [inlined]
│  [9] listenpkg(::getfield(Main, Symbol("##7#10")), ::Base.PkgId) at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:13
│  [10] top-level scope at C:\Users\kawcz\.julia\packages\Requires\9Jse8\src\require.jl:64
│  [11] top-level scope at C:\Users\kawcz\Dropbox\Caltech\work\dev\CLIMA\src\Utilities\test\runtests.jl:141
│  [12] include at .\boot.jl:317 [inlined]
│  [13] include_relative(::Module, ::String) at .\loading.jl:1038
│  [14] include(::Module, ::String) at .\sysimg.jl:29
│  [15] exec_options(::Base.JLOptions) at .\client.jl:229
│  [16] _start() at .\client.jl:421
│ in expression starting at C:\Users\kawcz\Dropbox\Caltech\work\dev\CLIMA\src\Utilities\test\runtests.jl:159
```

**Comment by charleskawczynski** [comment]
> I am not sure I understand the purpose of this PR.

@vchuravy,  I would like for `@testset "MoistThermo CUDA"` to pass.

> You do not want to directly use Cassette for this, GPUifyLoops is intended to abstract these things away for you so that you do not have to maintain this infrastructure.

I was following @lcw's slack comment:

> Actually I think you may be able to use contextualize even if the kernel is not using GPUifyLoops

So, to confirm, for `@testset "MoistThermo CUDA"` to pass, we should still use GPUifyLoops, correct?

File: src/Utilities/src/MoistThermodynamics.jl [hunk 2]
```diff
@@ -33,6 +35,14 @@ export liquid_fraction, phase_partitioning_eq!, saturation_adjustment
 # Auxiliary functions, e.g., for diagnostic purposes
 export liquid_ice_pottemp
 
+abstract type Device end
+abstract type  GPU end
+struct CPU <: Device end
+struct CUDA <: GPU end
+
+""" Allows contextualize to propagate GPU intrinsics """
+contextualize(::CUDA, f::F) where F = (args...) -> Cassette.overdub(ctx, f, args...)
```
[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/Utilities/src/MoistThermodynamics.jl]: I forgot that this is user defined.

**Comment by lcw** [comment]
I think I wasn't clear, I was suggesting to use
`GPUifyLoops.contextualize` even if you are not using ` GPUifyLoops.@loop`.

**Comment by vchuravy** [comment]
I plan to finish v0.0.2 of GPUifyLoops after which this should be a lot easier.

**Comment by charleskawczynski** [comment]
@asraero and I discussed #94, and it seems GPUifyLoops must be applied to the drivers, not MoistThermodynamics (which operates element-wise).

@vchuravy thanks for all of the updates on GPUifyLoops! The [example](https://github.com/vchuravy/GPUifyLoops.jl/blob/master/examples/contextualize.jl) you provided is a good MWE of what we need working.

I think that we can close this PR, or change the title to "GPUify the drivers" or something. @lcw @jkozdon what do you guys think about this?

**Comment by jkozdon** [comment]
Correct. We now cassette the whole kernel, so nothing special is required outside of the balance law solver.

We're working on GPUifying the balance law solver right now, so I think just let things come in with that.

**Comment by jkozdon** [comment]
Follow the [jek/BLprofileharness](https://github.com/climate-machine/CLIMA/tree/jek/BLprofileharness) branch if you want to see what's going on.

**Pull request closed without merging**